### PR TITLE
Rework features documentation

### PR DESCRIPTION
**Pull request by michaelpj**
REMINDER: RTD provides a preview of the rendered documentation via a PR status.

This does a few things:
- Adds a clear listing of which LSP methods we support, and which we
don't (requested a couple of times recently).
- Lists every code action and code lens, which plugin provies them,
their status, and code action kind where applicable (apart from Wingman
because there are lots and I couldn't figure it all out).

Possibly some of this information belongs elsewhere, e.g. perhaps
instead of listing which versions don't have full support on this page,
we should instead add a page to the GHC support page that lists which
plugins work on which versions. Not sure.

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2603"></a>

### CODE REVIEW

**Comment by michaelpj** [comment]
Also I would like to encourage plugin authors to provide slightly more substantial feature descriptions than my one-sentence summaries :sweat_smile:

**Comment by michaelpj** [comment]
Possible changes I thought about:
- GHC version compatibility in a table on the GHC support page instead?
- Code action kinds for each code action in a table somewhere instead of spread out?

**Comment by michaelpj** [comment]
Hmm, so maybe it would help to call out that you may need to consult your client to find out how to actually trigger these features?

**Comment by jneira** [comment]
there is an issue in this repo asking for an improvement of the section changed here: https://github.com/haskell/haskell-language-server/issues/2327
not sure if the pr fullfills some of the suggestions done there (or if it could do it)

**Comment by michaelpj** [comment]
I'm not convinced by the editor table, I think users should just consult the documentation for the editor, but otherwise I think this gets us a lot closer. Known limitations would be great... I just don't know the limitations in most cases. I think we'd need some help from plugin authors to make this significantly better.

**Review by jneira — APPROVED** [review]
great work as usual, I like the actual version but the alternatives sounds good as well

**Comment by michaelpj** [comment]
Ah yeah, my concern with the GHC version support stuff being spread out is that it's harder to maintain. Which is maybe the killer argument. This information is no good if it's out-of-date, and the GHC version support page is naturally the place we're going to think to update when we do GHC version support stuff. So for that reason I think it should go there, maybe with a link from here.

**Comment by GuillaumeDesforges** [comment]
Looks nice!

I've tried using Ctrl+F on some simple things I'd look for, like "hlint" or "explicit", and I got the info right away 👍 
I especially like the "missing features" part, which helps quickly figuring out why something can't be done (e.g. renaming).

File: docs/features.md
Comment (jneira) [inline docs/features.md:260]: does it worth mention the wip pr #2565?

Comment (michaelpj) [reply docs/features.md:260]: Eh, I'd wait until it's merged. It might never be, who knows!

**Comment by michaelpj** [comment]
- Moved plugin version support table to GHC support page
- Added a few known limitations
- Removed fairly pointless "Status" column from main feature table

**Comment by Anton-Latukha** [comment]
Personally, I think that in HLS docs it is sound to have plugin API info & examples how to develop them. Maybe include info/statuses on the closes built-in plugins, but not on all of them.

And lists of plugins, their description & status updates seem like should be in a separate repo. A repo that becomes a db on plugins & their support statuses.
That would allow to not get notifications & to not review the plugin list status update PRs.

Also woul'd be good if plugins could themselves tell which thing they support. That again goes to the coupling, if plugins are bundles - to tell is it supported or not - the closest people who have current updates on that info become core team responcibility.

What do you think on this?

**Comment by michaelpj** [comment]
> 1) plugin API info 2) examples on how to develop them

I don't think either of those are appropriate for a features page aimed at users of HLS, though?

As for the rest... the point of this PR is just to make the features page better, not redo the entire way that plugins are organized and documented!

**Comment by eddiemundo** [comment]
Some of the plugins have more substantial READMEs, should we inline some or most of that information here?

**Comment by michaelpj** [comment]
> Some of the plugins have more substantial READMEs, should we inline some or most of that information here?

Maybe. Probably some! I think we should keep this page relatively lightweight, and maybe if there are plugins that have lots to say (e.g. I'm sure there is lots to say about Wingman), we could add a documentation page specifically for them.

**Comment by eddiemundo** [comment]
Ah OK I'm thinking now we could just link to the readmes from this page if they have any.

File: docs/features.md
Comment (eddiemundo) [inline docs/features.md:141]:
```suggestion
For usage see the ![readme](plugins/hls-qualify-imported-names-plugin/README.md).

```

File: docs/features.md
Comment (eddiemundo) [inline docs/features.md:141]:
```suggestion

For usage see the ![readme](plugins/hls-qualify-imported-names-plugin/README.md).
```

File: docs/features.md
Comment (eddiemundo) [inline docs/features.md:144]:
```suggestion
```

File: docs/features.md
Comment (eddiemundo) [inline docs/features.md:144]:
```suggestion
```

File: docs/features.md
Comment (eddiemundo) [inline docs/features.md:142]:
```suggestion
For usage see the ![readme](/plugins/hls-qualify-imported-names-plugin/README.md).
```

File: docs/features.md
Comment (eddiemundo) [inline docs/features.md:142]:
```suggestion
For usage see the ![readme](../plugins/hls-qualify-imported-names-plugin/README.md).
```

**Comment by malteneuss** [comment]
> someone in the vscode extension asked recently for info like this (also key bindings but that is totally editor specific)
> haskell/vscode-haskell#525

Regarding documentation of key-bindings in HLS clients/editors, which is important for end-users, i'd like to propose the following idea as some future improvement to minimize, repeated manual work:

Could we mold the feature information into structured data, and then generate this page/table based on that?
Basically, provide (LSP Protocoll functions <-> Descriptions, Status, other Info). Then a client like vscode can provide (keybindings in vscode <-> LSP Protocoll functions), and combine both to generate a (keybindings in vscode <-> Descriptions, Status, other Info) Table on their documentation page.

That could be repeated for every HLS client in other editors.

**Comment by GuillaumeDesforges** [comment]
Would it be possible to add a link to each plugin's homepage, that that users can go learn more about them?

**Comment by michaelpj** [comment]
> Would it be possible to add a link to each plugin's homepage, so that users can go learn more about them?

@eddiemundo has been doing that, which is great. At the moment the existence of these is a bit spotty. I'd like to make them a more "proper" part of the documentation build, but that can happen later.

> Could we mold the feature information into structured data, and then generate this page/table based on that?
Basically, provide (LSP Protocoll functions <-> Descriptions, Status, other Info). Then a client like vscode can provide (keybindings in vscode <-> LSP Protocoll functions), and combine both to generate a (keybindings in vscode <-> Descriptions, Status, other Info) Table on their documentation page.

I guess my feeling is that the point of the LSP is that you get to do #clients + #servers work rather than #clients*#servers, and so I kind of want the same for docs: let the user put the server doc and the client doc together and do the "join" themselves. But then that is less user friendly.

Also we *can't* give keybindings for emacs in general since they're mapped to all sorts of things by various emacs distributions!

Anyway, a few additional requests in this thread, I'll make some issues for them so they don't get lost.

**Comment by michaelpj** [comment]
Better plugin doc: https://github.com/haskell/haskell-language-server/issues/2066
Keybindings: https://github.com/haskell/haskell-language-server/issues/2613

File: docs/supported-versions.md
Comment (Anton-Latukha) [inline docs/supported-versions.md:31]: Mention that it is GHC versions.

Also stating in HLS docs in article "Supported versions" that some plugins correspond to some Unsupported GHC versions - it raises a question, it is HLS not supports them for this GHCs?

"Additionally, some plugins are not supported on some GHC versions, as shown in the following table." only magnifies that belief: plugins are really are unsupported by HLS for particular versions. When the implication is in a different direction. It is plugins that do not support new GHCs & HLS releases that support those GHCs.

Stating: "Bounds on the support of GHC releases" & declaring the bounds below as Cabal bounds - would align the knowledge & understanding.

The first culprit of a particular plugin - is that it needs to support new GHC, which means support HLS API for that release, which means person goes - looks at `.cabal` description of the plugin in the mainline branch, clones & tweaks the boundary & observes the result & starts from there, & goes lurking on according reports/PRs.

**Comment by Anton-Latukha** [comment]
https://github.com/haskell/haskell-language-server/pull/2603#issuecomment-1016797883 well, sometimes a typo changes the reading.

The main point there - is that, as I think of it, plugins better to not be included into the documentation of HLS at all, only info on how to develop them & have "for additional features/to expend HLS - use plugins: <link to the plugin catalog>".

---

https://github.com/haskell/haskell-language-server/pull/2603#issuecomment-1016826913

> As for the rest... the point of this PR is just to make the features page better, not redo the entire way that plugins are organized and documented!

Nothing is more permanent than the temporary. Starting to catalog feature-by-feature catalog plugin abilities on `Features` page, means that users would grow to expect that all feature information cataloged there in this form, and that in short order plugins & users would start opening PRs to add their info into this article, which means permanent maintenance of the page, until design changes.

Besides even the scope of `haskell-language-server` is anthropologically perceived as development, not documentation. Because of the code development scope, people tend to treat repo as a code development repository and so completely forget about documentation.

"Features" page can be useful to state the basic functionality or the core projects of HLS & it is a great way to empirically transmit the structural knowledge/understanding. It is one thing to put in words what `ghcide` is & the other to know domain of user interface things that it does, knowing that it handles so and so - means that this code processing is present & the agenda of `ghcide` & so agenda aligns and code is in there & improvements to it would be accepted. It is overall also important that it forms the overall understanding of what `ghcide` does when to look from the user, which deepens the HLS stack understanding. But when all features are mentioned & `ghcide` is mentioned between a lot of plugins - that mixes-up the understanding & so `ghcide` agenda, scope & functionality as a whole stays not understood by the reader.

I do not want to draw the process in my thoughts. Just wanted to drive attention to the points that nothing is more permanent than the temporary, and that design decisions and maintenance required by that design matters.

Just giving an observation.

File: docs/features.md
Comment (Anton-Latukha) [inline docs/features.md:116]: This mentioning then probably needs to mention the caveat, that during doing that (for sure during the "apply all hints" - it frequently deletes CPP code blocks across the whole file.

Comment (Anton-Latukha) [reply docs/features.md:116]: Sorry, this may be out of date, because it appears I still use:
```
hlint --version
HLint v3.2.7
```
While there is `3.3.6`.

**Comment by malteneuss** [comment]
> I guess my feeling is that the point of the LSP is that you get to do #clients + #servers work rather than #clients*#servers, and so I kind of want the same for docs: let the user put the server doc and the client doc together and do the "join" themselves. But then that is less user friendly.

"#clients + #servers work .. for docs" is a nice summary what i had in mind. My proposal is to not let the users "join" that info (as you said as well); let the HLS clients do that (let them worry about their keybindings). But possibly we can make their join work much easier by agreeing on and providing features info as structured data, so that the join can be automated on their side.

**Comment by pepeiborra** [comment]
This is super-awesome, thank you @michaelpj

File: docs/supported-versions.md
Comment (michaelpj) [reply docs/supported-versions.md:31]: I'm sorry, I'm not sure I quite understand what you're saying. I will try and make the text clearer so that it's obvious that the lack of support is because the _plugin_ lacks support for that version of GHC.

**Comment by michaelpj** [comment]
> so users would grow to expect that all plugin feature information cataloged there in this form, and that in short order plugins & users would start opening PRs to add their info into this article, which means permanent maintenance of the page, until design changes.

That all sounds pretty good to me. At the moment we have basically no content about these things, so having some is a distinct improvement.

> But when all features are mentioned & ghcide is mentioned between a lot of plugins - that prevents to empirically infer the ghcide understanding & so scope & functionality of it stay not understood by the reader.

I think we still have a disconnect about who the _audience_ of this page is. I think the audience of this page is end-users, i.e. people who _use_ HLS but don't contribute. Understanding the structure of HLS is of no use to them, but knowing that a feature is associated with `ghcide` might be of some use.

Explaining the structure can be usefully done as part of the _contributing_ docs, IMO.

File: docs/features.md
Comment (michaelpj) [reply docs/features.md:116]: No, this is an `apply-refact` limitation AFAICT, and good to include. I've added it.

**Comment by michaelpj** [comment]
Okay, I think this is good enough state now.

File: docs/supported-versions.md
Comment (Anton-Latukha) [reply docs/supported-versions.md]: Yep.

Despite using English, sometimes I still do basic errors, like making a question while sequencing the words as a statement.

There is still a structural question, between HLS-plugin & GHC, there is a major structural part - HLS, but the table directs to compatibility with GHC. But that does not matter, users/contributors would raise the question if they would need it.

**Pull request merged**
